nss: prevent PROTECT_ERRNO from squashing changes to *errnop

yrro opened the pull request:
glibc passes in &errno for errnop, which means PROTECT_ERRNO ends up
squashing our intentional changes to *errnop.

This is a quick hack that adds a new flag alongside the saved value of
errno, which can be used to disarm PROTECT_ERRNO in the case that we
really do want to squash errno afterall.

Fixes #11321.

yuwata commented:
How about the following:
```
static inline void _reset_errno_(int *saved_errno) {
        If (*saved_errno >= 0)
		        errno = *saved_errno;
}
```
And set -1 when we want to disable resetting errno.

yuwata commented on src/basic/util.h:
```diff
@@ -173,12 +173,19 @@ static inline void *mempset(void *s, int c, size_t n) {
-        errno = *saved_errno;
+struct _protect_errno_data {
+        int value;
+        int enabled;
```
Please use `bool`.

yuwata commented on src/basic/util.h:
```diff
@@ -173,12 +173,19 @@ static inline void *mempset(void *s, int c, size_t n) {
+static inline void _reset_errno_(struct _protect_errno *p) {
+        if (p->enabled) {
+                errno = p->value;
+        }
```
Please drop unnecessary `{}`.

yrro commented:
I don't mind the change. But I wonder if anyone has feedback about the approach in general?

It seems that we're only in this situation because glibc happens to pass `&errno` as the value of `errnop`; my change bakes this assumption into `nss-resolve` which, as it is, quite reasonably believes that `errno` and `*errnop` are separate.

yuwata commented:
> But I wonder if anyone has feedback about the approach in general?

Yeah, let's wait other reviewers.

yrro commented:
One alternative would be to introduce `MAYBE_PROTECT_ERRNO` which would take an `int*`, and then modify the cleanup function to check that the value passed isn't equal to &errno before it restores the value. But that isn't thread-safe, and maybe it's just crazy to have this complicated mechanism just on the off chance that the client set this one pointer to a particular value... :)

yuwata commented:
After re-reading https://www.gnu.org/software/libc/manual/html_node/NSS-Modules-Interface.html and the discussion in #9587, now I think we should reset `errno` for the case that `errnop != &errno`.

So, how about replacing `PROTECT_ERRNO_DISARM` by the following:
```
#define PROTECT_ERRNO_FLUSH \
        errno = _saved_errno_; \
        _saved_errno_ = -1
```
By using this, `errno` is protected if `errnop != &errno`. Of course, `errno` is overridden if `errnop == &errno`, but it is a caller's problem.

@codonell Could you comment on this?

yuwata commented:
BTW, `test-fs-util` fails...

codonell commented:
> After re-reading https://www.gnu.org/software/libc/manual/html_node/NSS-Modules-Interface.html and the discussion in #9587, now I think we should reset `errno` for the case that `errnop != &errno`.
> 
> So, how about replacing `PROTECT_ERRNO_DISARM` by the following:
> 
> ```
> #define PROTECT_ERRNO_FLUSH \
>         errno = _saved_errno_; \
>         _saved_errno_ = -1
> ```
> 
> By using this, `errno` is protected if `errnop != &errno`. Of course, `errno` is overridden if `errnop == &errno`, but it is a caller's problem.
> 
> @codonell Could you comment on this?

I haven't read the code in systemd, but all the same principles apply as outlined in #9587. You can safely save and restore errno if you like, this will hide errors internal to your function, but that's OK, that's your own contract. Can you expand a bit more about what is going wrong and what you are trying to fix?

yuwata commented:
@codonell Thanks. The problem is that, when passed `errnop` is equivalent to `&errno`, then currently after setting proper value to `*errnop` we reset `errno`. So, the value assigned to `*errnop` vanishes.
Please see detail at #11321.

So, I proposed in the above comment that first reset `errno`, then assign value to `*errnop`.

yuwata commented:
@yrro More specifically, how about the following, as the macro will be used only in `nss-*`:
```
#define ASSIGN_ERRNOP(r) \
        ({ \
            errno = _saved_errno_; \
            _saved_errno_ = -1; \
            abs(r); \
        })
```

yrro commented:
Not 100% I'm following what you mean, but how about this?

Note that I'm not sure what to do in the code labelled `fail`, since that assigns a negative value to *errnop.

Is the behaviour correct in statically-linked programs (where `*errnop != &errno`)? Or should `ASSIGN_ERRNOP` only assign to `errno` and `_saved_errno_` if they point to the same address?

poettering commented:
@yrro hmm, could you squash your patches into one? we generally want to merge "perfect" PRs, i.e. where commits reflect logical steps, but not historical ones. i.e. in this case it appears to me later commits fix up previous commits of the same PR, and that's what we generally don't want, to keep things bisectable. It's also harder to review if you keep looking at commits that are completely reworked later on!

poettering commented:
I think I like the current final version in this PR mostly, but I'd like to ask you for two changes: can you move the macro definition to util.h next to PROTECT_ERRNO's definition, please? Also, while I think the macro's definition is good I'd like it to be renamed to "DISARM_PROTECT_ERRNO" or DISARM_ERRNO_PROTECTION" or "UNPROTECT_ERRNO" or so... I'd also prefer if there'd be a brief comment there that explains why one would ever want to use this macro, with a clear reference to NSS...

poettering commented:
> @poettering Yup, I'll squash everything down once I'm clear what the best direction for the PR is.
> 
> Just to be clear, do you mean rename `ASSIGN_ERRNOP` so that clients will use it like:
> 
> `*errnop = DISARM_PROTECT_ERRNO(ERANGE)`

yupp!

yrro commented:
(Looks like I'm not the first person to ever wonder about this design... https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=468378)

yuwata commented:
LGTM. But, I think the other NSS modules, nss-mymachines, nss-myhostname, and nss-systemd should be also updated.

yuwata commented on src/nss-resolve/nss-resolve.c:
```diff
@@ -267,7 +267,7 @@ enum nss_status _nss_resolve_gethostbyname4_r(
 
 fail:
-        *errnop = -r;
+        *errnop = DISARM_PROTECT_ERRNO(-r);
```
Not a matter, but please drop minus here, as the macro contains `abs()`.

poettering replied:
please drop the "-", it's unnecessary now

yuwata commented on src/nss-resolve/nss-resolve.c:
```diff
@@ -455,7 +455,7 @@ enum nss_status _nss_resolve_gethostbyname3_r(
 
 fail:
-        *errnop = -r;
+        *errnop = DISARM_PROTECT_ERRNO(-r);
```
Also here.

yuwata commented on src/nss-resolve/nss-resolve.c:
```diff
@@ -636,7 +636,7 @@ enum nss_status _nss_resolve_gethostbyaddr2_r(
 
 fail:
-        *errnop = -r;
+        *errnop = DISARM_PROTECT_ERRNO(-r);
```
Also here.

poettering commented:
> Ok great. Do you have any comments about what should happen when the NSS modules are invoked by a statically-linked program? In this case `&errno` and `errnop` will be different and the NSS module should modify only *errnop.

i am not sure how that could ever work... our NSS modules are linked against libc.so, so you are saying you then end up with the whole libc twice in memory? how can that ever work? i mean there are plenty APIs (such as futexes) which can only have a single user-space consumer, so how could you ever mix two libcs in memory for the same process? quite frankly, "compiling things statically" and "using NSS" are two contradicting things to do, afaiu. @codonell is something like that something you intend to support with glibc?

but regardless, if this is an intended usecase of glibc and the NSS, and there indeed is a scheme how you can load NSS modules dynamically into statically linked processes, and you thus have two errno instances, why would this mean anything special for our nss modules? i mean, we will only change the errno in our copy, and we will reset it again to its original value in the end. If that instance is shared with the rest of the process or not doesn't really matter. in the worst case we restore a value unnecessarily, but note that the whole restoring of errno is an unnecessary excercise anyway, as it is just a work-around for old glibc and unnecessary on newer glibc.

poettering commented on src/basic/util.h:
```diff
@@ -174,12 +174,26 @@ static inline void *mempset(void *s, int c, size_t n) {
+ */
+#define DISARM_PROTECT_ERRNO(r) \
+        ({ \
+                errno = _saved_errno_; \
```
please make this idempotent, i.e. if _saved_errno_ is already negative don't restore errno here either

codonell commented:
@poettering Statically linked applications have a static namespace which is distinct from the dynamic namespace (and yes everything gets loaded twice, and we remind people of this via the warning " warning: Using 'dlopen' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking") of dynamically loaded modules. We warn about this very thing in the C/C++ guidelines for Fedora (https://fedoraproject.org/wiki/C_and_C%2B%2B_v2#Static_Linking). There are indeed many things that do not work in static applications. Upstream supports static linking with caveats. We are actively trying to make the NSS modules interface into a kind of service that would avoid dynamic linking (https://www.sourceware.org/ml/libc-alpha/2017-12/msg00831.html, https://sourceware.org/ml/libc-alpha/2017-12/msg00521.html). The NSS modules are dynamically loaded and therefore occupy a dynamic namespace within the static application, and they have their own unique copy of errno which is not seen by the caller (the caller resides in the static namespace). This is why we have a pointer to errno to allow the dynamically loaded NSS modules to write into the caller's errno location. In the case of statically linked applications the NSS module failures cease to propagate via errno at the boundary of the namespace. I hope that answers all of your questions.

poettering commented:
@codonell hmm, but iirc futexes and stuff can only have one consumer in userspace per process. So if you have two copies of libc in memory for the same processes, what happens to them? which libc instance owns them?

codonell commented:
> @codonell hmm, but iirc futexes and stuff can only have one consumer in userspace per process. So if you have two copies of libc in memory for the same processes, what happens to them? which libc instance owns them?

I'm not sure I know how to answer your question without a concrete example, but I will try to give you some more background. In practice each libc instance has an entirely unique set of memory for use with data and TLS, and so any futexes using that memory are entirely distinct. The worst case scenario is threads which have only one thread register and so one way to access TLS data, but even in that case the second dynamically loaded libc is given it's own unique block of TLS data. As you surmise this may lead to things appearing "locked" in the static namespace and "unlocked" in the dynamic namespace, and this is exactly why we do not recommend using any threaded applications with static linking (you generally immediately run into issues with locales first).

poettering commented:
> BTW, presently a statically-linked program segfaults when it calls into `libnss_resolve`:

this appears to segfault in a c11 "thread_local" variable. My educated guess is that thread local variables are not compatible with this kind of "static binary with dynamically loaded module" mode, as most likely the thread local vars of the dynamic module are never allocated properly for threads created in the static libc... @codonell any idea? but anyway, this is a different issue and not sure we can do much about it from the systemd side... Quite frankly, static binaries are more theory than practical, not sure its worth the effort your are putting into them, but that's of course up to you...

codonell commented:
> > BTW, presently a statically-linked program segfaults when it calls into `libnss_resolve`:
> 
> this appears to segfault in a c11 "thread_local" variable. My educated guess is that thread local variables are not compatible with this kind of "static binary with dynamically loaded module" mode, as most likely the thread local vars of the dynamic module are never allocated properly for threads created in the static libc... @codonell any idea? but anyway, this is a different issue and not sure we can do much about it from the systemd side... Quite frankly, static binaries are more theory than practical, not sure its worth the effort your are putting into them, but that's of course up to you...

Static linking is really designed for single-threaded applications that need to do one thing in a constrained environment without relying on the OS or available libraries to carry out their task. If you can limit yourself in this regard then you can run static applications like this, and we try to give this choice to users. The problem over the years has been users expecting more and more of the runtime to work in a statically compiled scenario, and honestly we've not done a great job of educating our users about the limits. In the case of Red Hat Enterprise Linux for example we officially do not support static linking. Likewise in Fedora we have strong language and explicitly call out that if you create threads in a statically linked application, and then call functions that require NSS, that this is very likely to fail. The exact causes of the failure vary, in the case of "thread_local" there is an underlying runtime registration interface in the C library that has to handle the destructor registration for non-POD objects, and it may be that interface is not quite setup. I have not verified this exact failure mode.

yrro commented:
Based on the discussion above I'll just not bother to think too hard about the static linking case then. :)

`DISARM_PROTECT_ERRNO` can't be used in `fill_in_hostent` because `_saved_errno_` is not accessible. I've modified it to take a pointer to _saved_errno_ and introduced `DISARM_PROTECT_ERRNO_INNER` which works via a pointer.

keszybz commented:
Looks like all comments have been addressed. Let's merge.

yrro commented:
Thanks folks!
